### DISCUSSION TITLE
Feature Request: User-defined Higher-Order Objects

### DISCUSSION THREAD

**Discussion by rtbs-dev** [InvestmentSystems/static-frame#709] [Ideas]
**Description**

Static-frame nicely handles the typical use-cases of higher-order frame-like objects. However, there are a few communities that deal in trying to make more specific higher-order objects that achieve specific ends. 

This feature request is to provide a mechanism for users to build their own higher-order containers, based on a similar idea to what is behind Bus, Yarn, Quilt, etc. If this functionality already exists, perhaps a tutorial of using the static-frame interface types for building their own "meta"-containers would be enough. 

Why this is useful: there are a number of "hierarchical data format" providers, mostly centered around HDF5. This includes Xarray, AnnData, Muon, etc. Many of these take a very object-oriented approach, but static-frames provides a way to do this with a more immutable, functional-driven approach. The existing higher-order containers provide transparent layers of abstraction, but would (presumably) need extending for some of these use-cases (multi-axis alignment, etc.)
 
**Example**

See this [anndata issue](https://github.com/scverse/anndata/issues/742) where static-frames is mentioned. Note that a "meta" object (a dataclass) is created to represent a set of static frames (or sparse arrays) that are aligned on one of two axes. the "X" object is a frame aligned to _both_ (observations index, features index), and there is a (in static-frame terminology) "batch" of frames called "layers" also aligned to look like X that can contain other values. 

Here's a graphic from [their docs](https://anndata.readthedocs.io/en/latest/): 
![anndata architecture](https://anndata.readthedocs.io/en/latest/_images/anndata_schema.svg)

Note that there's a section in their docs that describes a way to lazily concatenate a bunch of these into "one", which is incredibly reminiscent of a `yarn`, but I don't see a way to build such a consistent yarn of this specific heterogeneous structure. Then, if we dig a bit more, an abstraction limitation in the `anndata` framework led to an entire new `muon` system that acts like a [dictionary of anndata](https://muon.readthedocs.io/en/latest/io/mudata.html). See [their paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02577-8) if interested.  XArray may allow a similar system, but in essence we are talking about things that serialize to HDF5 in its more general "file-folder-like" form. 

**What would this look like?**
Static-frame currently allows for higher order containers of one axis. Supporting more general higher-order systems would mean heterogeneously indexed containers. My interest in this functionality is effectively creating flexible systems for _Incidence Structures_: 

$$(S_1, S_2, I), \text{s.t.} I\in S_1\times S_2$$

Here, S1 and S2 are their own "frames" (which could be Batches loaded lazily) containing metadata for every member, and I is either a series with a IndexHierarchy made from subsets of (S1, S2) to make a sparse array, or a frame indexed as S1 and columns S2. 

Then, I would like to lazily load/store Buses (or...quilts?) of these heterogeneously indexed objects. 

To take this even higher, I would then like to make a container of these representing pairwise incidences of many different types of sets $S_1,S_2,...S_k$. This is exactly what the [Rank-k Structures](https://www.wikiwand.com/en/Incidence_structure#Generalization) represent, and is what Muon is creating, as far as I can tell. 

I can imagine a couple ways this might work: 

- Document the way the current higher-order containers are made, such that users could build custom containers, presuming the source code is amenable to this kind of extending. 
- Make an extension/plugin api for users to define them, allowing deeper interop with libraries like XArray, anndata, muon, loompy, etc. This would actually look a lot like letting users build custom SF containers that map onto arbitrary Zarr or HDF/PyTables structures. Plus there's the lovely transparency in HLoc and other lazy operations we have in static frame by default. 
- create static-frame interfaces to these other libraries via, e.g. the numpy protocol, letting them "contain" static frames. This could work, but also may run counter to the pre-existing higher-order container system. I'd prefer nice parsing between libraries, over needing to depend on two libraries and a plugin between them. 
- I might misunderstand the current containers, and all of this could already be possible. If so, i'd be happy to contribute a tutorial or example illustrating how static-frame can accomplish such things to work with the other libraries I mentioned.

**Comment by flexatone** [comment]
Many thanks for your detailed post and an introduction to AnnData and Muon. I am happy to hear of your interest in StaticFrame's higher-order containers too.

I fully support increasing the interoperability between StaticFrame and other libraries, and StaticFrame containers already support some aspects of the NumPy array protocols (though you generally need to use the ``Frame.via_values`` interface, which could be expanded). If you have specific priorities (such as Zarr or Loom support) that is something I would be happy to explore.

User-defined custom containers built on StaticFrame's containers is a fascinating idea, but I think I need a more concrete example of what you are looking for to get a better sense of how to approach it. Two obvious options are composition (a new container that holds, for example, many ``Bus`` and enforces certain cross-container alignments) or inheritance (subclassing a ``Bus`` or other container and implementing additional features). My impression is that composition might be the way to go for something like what you describe as an Incidence Structure. Or  there might be an entirely new higher-order container needed.

Can you describe a specific example of the the minimal viable functionality for such containers?

Again, thanks for this post and your ideas and I look forward to learning more.
